### PR TITLE
CJAUnit 5 - Firebase Realtime Database Implementation

### PR DESCRIPTION
**Pull request by frankiebarrios**
## What It Does
This pull request is set to complete mini-project: add a database of unit 5 of the ux-onboarding-project.

This work is taking place to complete mini-project: add a database of unit 5 of the ux-onboarding-project.

**Implementation:**
- Installed, configured, and implemented Firebase Realtime Database.
- Rewrote LocalStorage API to utilize Firebase Database.

## How To Test
- Install dependencies using instruction in README.
- Start server using yarn start or build using yarn build.

## Notes/Caveats

None
## Checklist

Under penalty of public shaming, I avow that I:
**Many items in list N/A** 
- [X] Reviewed the diff to look for typos, whitespace errors, and console.log() statements
- [X] Verified that there are no compiler or linter errors or warnings
- [X] Verified the build in production(optimized) mode
- [X] Updated the docs, if necessary
- [X] Included the appropriate labels
- [X] Referenced applicable issues (i.e. Closes #XXXX, Fixes #XXXX)

Browsers tested:
**Many items in list N/A** 
- [X] Chrome
- [X] Firefox
- [ ] Safari
- [ ] Edge
- [ ] Internet Explorer

## Dependencies
1. Install Node and npm
2. Install Yarn
3. Install Ux-Lint

### CODE REVIEW

File: src/api/firebaseAPI.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    const user = this.getUser(id);
+    const ref = firebase.database().ref('users');
+    let match = '';
+    ref
+      .orderByChild('id')
+      .equalTo(user.id)
+      .on('child_added', function (snapshot) {
+        match = snapshot.key
+      });
+    firebase
```
Comment (kddrake) [inline src/api/firebaseAPI.js:85]: what happens here if a match isn't found?

File: src/api/firebaseAPI.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    ref.on('value', this.getData.bind(this));
+    return this.users;
+  }
+
+  getData (data) {
+    try {
+      const firebaseData = data.val();
+      const keys = Object.keys(firebaseData);
+      const userObj = [];
+      for (let i = 0; i < keys.length; i++) {
```
Comment (kddrake) [inline src/api/firebaseAPI.js]: this for could be simplified
```suggestion
      keys.forEach(key => { userObj.push(firebaseData[key]) });
```

File: src/api/firebaseAPI.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        let key = keys[i];
+        let user = firebaseData[key];
+        userObj.push(user);
+      }
+      this.users = userObj;
+    } catch (error) {
+      console.log('Error Pulling In Users: ', error);
```
Comment (kddrake) [inline src/api/firebaseAPI.js]: if you're going to log errors to the console use `console.error()`

File: src/components/intern-app.html [hunk 4]
```diff
@@ -3,6 +3,10 @@
 <link rel="import" href="./user-list.html">
 <link rel="import" href="./edit-user.html">
 <link rel="import" href="./create-user.html">
+<!-- The core Firebase JS SDK is always required and must be listed first -->
+<script src="https://www.gstatic.com/firebasejs/6.4.0/firebase-app.js"></script>
+<!-- Firebase Database Link -->
+<script src="https://www.gstatic.com/firebasejs/5.0.1/firebase-database.js"></script>
```
Comment (kddrake) [inline src/components/intern-app.html]: you should try adding these your your `package.json` over in-line scripting. I assume it's possible, though I don't fully know for sure. I think Christa was able to.

**Review by kddrake — CHANGES_REQUESTED** [review]

File: src/components/intern-app.html (hunk 4 above)
Comment (frankiebarrios) [reply src/components/intern-app.html]: Turned out I didn't need that at all. Removed it.

File: src/components/intern-app.html [hunk 5]
```diff
@@ -51,25 +65,21 @@
...
-      ready() {
-        super.ready();
         this.addEventListener('updateUserList', this.updateUserList);
       }
-      
-      updateUserList(e) {
-        this.users = this.storage.getAllUsers();
+
+      updateUserList() {
+        this.users = this.storage.getAllUsers();
```
Comment (kddrake) [inline src/components/intern-app.html]: get rid of this trailing whitespace

File: src/components/create-user.html [hunk 6]
```diff
@@ -115,18 +124,79 @@ <h1 class="div-center">Create User</h1>
...
+          this.storage.createUser(this.user);
+        }
+        this.hasError = false;
+        this.dispatchEvent(new CustomEvent('updateUserList', { bubbles: true, composed: true }));
       }
 
-      updateUserList() {
-        this.createUser();
+      deleteAllUsers() {
+        this.storage.deleteAllUsers();
```
Comment (kddrake) [inline src/components/create-user.html]: lets get rid of this trailing whitespace here, too

**Review by kddrake — CHANGES_REQUESTED** [review]
The validation work in here should really be in it's own PR (or in a friction clean up PR) as it doesn't really have anything to do with your database.

**Review by kddrake — CHANGES_REQUESTED** [review]
run through it in the browser and check your console, seeing a couple of errors. Edit isn't working, either.

File: src/components/intern-app.html [hunk 7]
```diff
@@ -51,25 +65,21 @@
...
```
Comment (kddrake) [inline src/components/intern-app.html]: need to remove this trailing whitespace

**Review by kddrake — CHANGES_REQUESTED** [review]
your create-user component looks like it has friction work in it. needs to be removed, or you need to convince me it needs to stay in.

File: src/components/create-user.html [hunk 8]
```diff
@@ -74,22 +72,25 @@
...
+          <input type="text" id="lastName" placeholder="Your last name.." value="{{user.lastName::input}}">
 
-          <label for="phone">Phone</label>
-          <input type="text" placeholder="Your phone number.." value="{{user.phone::input}}">
-
-          <label for="email">Email</label>
-          <input type="text" placeholder="Your email.." value="{{user.email::input}}">
+          <label>Phone</label>
+          <input type="text" id="phone" placeholder="Your phone number.." value="{{user.phone::input}}">
+
```
Comment (kddrake) [inline src/components/create-user.html]: still got some whitespace here that'll need removed.

**Review by kddrake — CHANGES_REQUESTED** [review]

**Review by kddrake — CHANGES_REQUESTED** [review]
Should probably pull the list of users on init. Kinda weird to have to click a button.

**Comment by frankiebarrios** [comment]
> Should probably pull the list of users on init. Kinda weird to have to click a button.

@kddrake Have made that change in the UI Friction branch.

**Review by kddrake — APPROVED** [review]

**Pull request merged**